esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -2194,52 +2194,41 @@ void CWsDfuEx::doGetFileDetails(IEspContext &context, IUserDescriptor* udesc, co
 }
 
 
-void CWsDfuEx::getLogicalFileAndDirectory(IEspContext &context, IUserDescriptor* udesc, const char *dirname, IArrayOf<IEspDFULogicalFile>& LogicalFiles, int& numFiles, int& numDirs)
+void CWsDfuEx::getLogicalFileAndDirectory(IEspContext &context, IUserDescriptor* udesc, const char *dirname,
+    bool includeSuperOwner, IArrayOf<IEspDFULogicalFile>& logicalFiles, int& numFiles, int& numDirs)
 {
     double version = context.getClientVersion();
     if (dirname && *dirname)
         PROGLOG("getLogicalFileAndDirectory: %s", dirname);
     else
         PROGLOG("getLogicalFileAndDirectory: folder not specified");
 
-    StringArray roxieClusterNames;
-    IArrayOf<IEspTpCluster> roxieclusters;
-    CTpWrapper dummy;
-    dummy.getClusterProcessList(eqRoxieCluster, roxieclusters);
-    ForEachItemIn(k, roxieclusters)
-    {
-        IEspTpCluster& cluster = roxieclusters.item(k);
-        StringBuffer sName = cluster.getName();
-        roxieClusterNames.append(sName.str());
-    }
-
     numFiles = 0;
     numDirs = 0;
     if (dirname && *dirname)
     {
-        StringBuffer filter;
-        filter.append(dirname);
-        filter.append("::*");
+        StringBuffer filterBuf;
+        setFileNameFilter(NULL, dirname, filterBuf);
+        if (includeSuperOwner)
+            filterBuf.append(DFUQFTincludeFileAttr).append(DFUQFilterSeparator).append(DFUQSFAOincludeSuperOwner).append(DFUQFilterSeparator);
 
-        Owned<IDFAttributesIterator> fi = queryDistributedFileDirectory().getDFAttributesIterator(filter.toLowerCase().str(), udesc, false,true, NULL);
-        if(fi)
-        {
-            StringBuffer size;
-            ForEach(*fi)
-            {
-                IPropertyTree &attr=fi->query();
+        //filters used to filter query result received from dali server.
+        DFUQResultField localFilters[8];
+        localFilters[0] = DFUQRFterm;
 
-                StringArray groups;
-                if (!getFileGroups(&attr,groups))
-                    groups.append("");
+        DFUQResultField sortOrder[2] = {DFUQRFname, DFUQRFterm};
 
-                ForEachItemIn(i, groups)
-                {
-                    addToLogicalFileList(attr, groups.item(i), version, LogicalFiles);
-                    numFiles++;
-                }
-            }
-        }
+        __int64 cacheHint = 0; //No page
+        unsigned totalFiles = 0;
+        bool allMatchingFilesReceived = true;
+        Owned<IDFAttributesIterator> it = queryDistributedFileDirectory().getLogicalFilesSorted(udesc, sortOrder, filterBuf.str(),
```

Comment:
I am using the getLogicalFilesSorted() to replace the getDFAttributesIterator() because the getDFAttributesIterator() does not return SuperOwner. Another option is to add SuperOwner to getDFAttributesIterator(). I did not choose that option because it may decrease the performance of getDFAttributesIterator() which is used in MANY other places. The SuperOwner was added to getLogicalFilesSorted() as an option in https://github.com/hpcc-systems/HPCC-Platform/pull/7206.